File: HW08/hw8/src/ArrayList.java
/**
 * ArrayList class.
 * @author Cihan
 * @param <E> the type of elements in this collection.
 */
public class ArrayList<E> implements List<E> {
	/**
	 * Array of elements E;
	 */
	private E[] elements;
	/**
	 * Size of array that includes element.
	 */
	private int size=0;
	
	private double sizeFactor = 0.75;
	
	/**
	 * No parameter constructor.
	 */
	@SuppressWarnings("unchecked")
	public ArrayList(){
		elements = (E[]) new Object[8];
	}
	/**
	 * One parameter constructor.
	 * @param capacity capacity of array.
	 */
	@SuppressWarnings("unchecked")
	public ArrayList(int capacity) {
		elements = (E[]) new Object[capacity];
	}
	
	@Override
	public MyIterator<E> iterator() {
		return new MyIterator<E>() {
			int index = 0;
			
			@Override
			public boolean hasNext() {
				if(index < size) {
					return true;
				}
				return false;
			}
			
			@Override
			public E next() {
				return elements[index++];
			}
			
			@Override
			public E previous() {
				return elements[index--];
			}
			
			@Override
			public int position() {
				return index;
			}
			
			@Override
			public void remove() {
				ArrayList.this.remove(elements[index]);
			}
		};
	}

	@Override
	public boolean add(E e) {
		MyIterator<E> myiter = this.iterator();
		
		while(myiter.hasNext()) { // Checking next.
			myiter.next();
		}
		
		if(shouldGrowArrayList(elements)) {	//Checking for size and growing up the set.
				growArray();
		}
		elements[size] = e;
		size++;
		
		return true;
	}

	@Override
	public boolean addAll(Collection<E> c) {
		MyIterator<E> otherIter = c.iterator();

		if(c.isEmpty()){
			return false;
		}

		while(otherIter.hasNext()){
			add(otherIter.next());
		}

		return true;
	}

	@Override
	public void clear() {
		size = 0;
	}

	@Override
	public boolean contains(E e) {
		MyIterator<E> iter = iterator();

		while(iter.hasNext()){
			if(iter.next()== e){
				return true;
			}
		}

		return false;
	}

	@Override
	public boolean containsAll(Collection<E> c) {
		MyIterator<E> otherIter = c.iterator();

		boolean res=true;

		while(otherIter.hasNext()){
			if(!(contains(otherIter.next()))){
				res = false;
			}
		}

		return res;
	}

	@Override
	public boolean isEmpty() {
		return size==0;
	}

	@Override
	public boolean remove(E e) {
		MyIterator<E> removeIt = this.iterator();
		
		boolean res = false;
		
		while(removeIt.hasNext()) {
			if(removeIt.next().equals(e)) {
				res = true;
				break;
			}
		}
		
		if(res) {
			for(int i=removeIt.position();i<size;i++) {
				elements[i-1] = elements[i];
			}
			size--;
		}
		return res;
	}

	@Override
	public boolean removeAll(Collection<E> c) {
		
		if(c.isEmpty()){
			return false;
		}
		boolean res=true;
		
		MyIterator<E> otherIter = c.iterator(); // iterator for incoming collection.

		while(otherIter.hasNext()){
			if(!(remove(otherIter.next()))){
				res=false;
			}
		}
		
		return res;
	}

	@Override
	public boolean retainAll(Collection<E> c) {
		MyIterator<E> myIter = iterator();
		HashSet<E> temp = new HashSet<E>();	 // creating a temp hashset.

		while(myIter.hasNext()){
			if( !(c.contains(myIter.next())) ){
				myIter.previous();
				temp.add(elements[myIter.position()]);
				myIter.next();
			}
		}

		if(removeAll(temp)) {
			return true;
		}
		return false;
	}

	@Override
	public int size() {
		return size;
	}
	
	
	/**
	 * Checks if the arraylist needs to grow.
	 * @param elementarray
	 * @return true if it need grow else false.
	 */
	private boolean shouldGrowArrayList(E[] elementarray) {
		if(size > elementarray.length * sizeFactor) {
			return true;
		}
		return false;
	}
	
	/**
	 * Growing up the arraylist.
	 */
	@SuppressWarnings("unchecked")
	private void growArray() {
		
		E[] newElements = (E[]) new Object[2*elements.length];
		
		for(int i = 0 ; i<size ; i++) {
			
			newElements[i] = elements[i];
			
		}
		
		elements = (E[]) new Object[2*elements.length];
		
		for(int i = 0 ; i<size ; i++) {
			
			elements[i] = newElements[i];
			
		}
	}
	

}

File: HW06/arraylist.cpp
#include "arraylist.h"

namespace MyCollection{
	
	template<typename E,typename container>
	// This function returns a container iterator for us.
	Iterator<E,container> ArrayList<E,container>::iterator()  {
		Iterator<E,container> iter(_c);
		return iter;
	}
	template<typename E,typename container>
	// Adds the element in this arraylist. Arraylist can have duplicate elements.
	bool ArrayList<E,container>::add(const E &e)   {
		Iterator<E,container> iter = iterator();

		for(int i=0; i<size();i++){
			iter.Next();
		}
		_c.insert(iter.getIteratorBegin(),e);
		return true;
	}
	template<typename E,typename container>
	// Adding all incoming collection's elements to our collection.If we can add all of them then returns true.Otherwise returns false.
	bool ArrayList<E,container>::addAll(Collection<E,container> &c)  {
		Iterator<E,container> otherIter = c.iterator();

		if(c.isEmpty()){
			return false;
		}

		while(otherIter.hasNext()){
			add(otherIter.Next());
		}

		return true;
	}
	template<typename E,typename container>
	//Clearing the container.
	void ArrayList<E,container>::clear()  {
		_c.clear();
	}

	template<typename E,typename container>
	//Returns true if this collection contains the specified element.
	bool ArrayList<E,container>::contains(const E &e)  {
		Iterator<E,container> iter = iterator();

		while(iter.hasNext()){
			if(iter.Next()== e){
				return true;
			}
		}

		return false;

	}
	template<typename E,typename container>
	//Returns true if this collection contains all of the elements in the specified collection.
	bool ArrayList<E,container>::containsAll(Collection<E,container> &c)  {

		Iterator<E,container> otherIter = c.iterator();

		bool res=true;

		
		while(otherIter.hasNext()){
			if(!(contains(otherIter.Next()))){
				res = false;
			}
		}

		return res;
	}
	template<typename E,typename container>
	//Returns true if this collection contains no elements
	bool ArrayList<E,container>::isEmpty()  {
		return _c.size() == 0;
	}
	template<typename E,typename container>
	//Removes a single instance of the specified element from this collection.
	void ArrayList<E,container>::remove(const E &e)  {

		Iterator<E,container> iter = iterator();

		if(*iter.getIteratorBegin() == e){
			_c.erase(iter.getIteratorBegin());
			return;
		}
		else{
			while(iter.hasNext()){
				if(iter.Next() == e){
					_c.erase(--iter.getIteratorBegin());
					return;
				}
			}
		}

	}
	template<typename E,typename container>
	//Removes all of this collection's elements that are also containedin the specified collection
	void ArrayList<E,container>::removeAll(Collection<E,container> &c)  {

		Iterator<E,container> otherIter = c.iterator();

		while(otherIter.hasNext()){
			remove(otherIter.Next());
		}

	}
	template<typename E,typename container>
	//Retains only the elements in this collection that are contained in the specified collection
	void ArrayList<E,container>::retainAll(Collection<E,container> &c)  {
		Iterator<E,container> myIter = iterator();
		ArrayList<E,container > temp;

		while(myIter.hasNext()){
			if( !(c.contains(myIter.Next())) ){
				--myIter;
				temp.add(myIter.Next());
			}
		}

		removeAll(temp);
	}
	template<typename E,typename container>
	//Returns the size.
	int ArrayList<E,container>::size()  {
		return _c.size();
	}

}
